### DISCUSSION TITLE
Proposal: support custom async iterator method builder

### DISCUSSION THREAD

**Discussion by neuecc** [dotnet/csharplang#3629] [Language Ideas]
`await foreach` supports any type that implements `GetAsyncEnumerator`.
But async iterator is tightly coupled `IAsyncEnumerable<T>` and `ManualResetValueTaskSourceCore`, that can not configure it.
Also `Task` has task-like custom async method builder feature.
My proposal is could you provides custom async iterator method builder?

For example:

```csharp
[AsyncIteratorMethodBuilder(typeof(UniTaskAsyncEnumerableBuilder)]
public interface IUniTaskAsyncEnumerable<out T>
{
    IUniTaskAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken = default);
}

public interface IUniTaskAsyncEnumerator<out T> : IUniTaskAsyncDisposable
{
    T Current { get; }
    UniTask<bool> MoveNextAsync();
}

public interface IUniTaskAsyncDisposable
{
    UniTask DisposeAsync();
}

public struct UniTaskAsyncEnumerableBuilder
{
    // like TaskMethodBuilder...
}

// usage
public async IUniTaskAsyncEnumerable<int> EveryUpdate([EnumeratorCancellation]CancellationToken cancelationToken)
{
    var frameCount = 0;
    await UniTask.Yield();
    while (!cancelationToken.IsCancellationRequested)
    {
        yield return frameCount++;
        await UniTask.Yield(); // next frame of game loop
    }
}
```

For the real usage, I've already provides custom async enumerable type(`IUniTaskAsyncEnumerable`, `IUniTaskAsyncEnumerator`) and async LINQ in my OSS [Cysharp/UniTask](https://github.com/Cysharp/UniTask/).
This is for the Unity Game Engine, and that custom Task(like ValueTask/IValueTaskSource) is tuned for the game engine.

**Comment by timcassell** [comment]
I was just looking at adding this to my Promise library, only to find out it's not possible. Now I see your proposal from a year ago!
